### ISSUE TITLE
Refactor to allow custom errors for custom formats

### ISSUE THREAD

**Issue by anilanar** [xeipuuv/gojsonschema#84]
As the title says, that would be nice.

**Comment by cristiangraz** [comment]
@anilanar are you looking for custom error message based on the specific format (so that not all format errors return the same string? If so, there is an `err.Details()` function that returns a `map[string]interface{}` and you can do something like this in your application:

```go
msg := err.Description()
if err.Type() == "format" {
   details := err.Details()
   switch details["format"] {
      case "formatA":
         msg = "Format A error"
      case "formatB":
         msg = "Format B error"
  }
}

return msg
```

IMO the library shouldn't provide that level of customization, it should return the errors and all of the detail needed and library users can convert the output to a custom format. Some users may want error strings, others might want custom structs with json or xml tags.

I'm using this library in an api "framework" and I have a validator function that validates and maps json schema errors from this library to a specific struct with defined Field/Type/Message fields that can be outputted to json or xml.

**Comment by anilanar** [comment]
@cristiangraz 

I've just checked the code written back in Nov 2015. I have some custom format validators. One of them:

```go
import (
	"regexp"

	"github.com/xeipuuv/gojsonschema"
)

var regex = regexp.MustCompile(`^(?=[A-Z])(?=[a-z])(?=[1-9]).{8,}$`)

type Password struct{}

func (p Password) IsFormat(input string) bool {
	return regex.FindStringIndex(input) != nil
}

func RegisterPasswordValidator() {
	gojsonschema.FormatCheckers.Add("password", Password{})
}
```

I guess I wished to be able to define error message on the struct, maybe by defining a `GetMessage` method? I don't really remember though.

**Comment by cristiangraz** [comment]
Yeah that's a good use for formats, and I can understand the need to want to pass along custom messages to your users.

There is an open issue #62 about converting format checkers to functions (which was just my oversight when I initially build out the format checker functionality). If custom formats _did_ support custom messages, it would be a good reason for using a struct and having an additional method to meet the interface for defining a formatter with it's own validator (`IsFormat`) and it's own message (`GetMessage` maybe).

I'm not a library maintainer so not my call, but personally think it's beyond the scope of the library which IMO should just be to validate JSON schema and provide detailed errors back.
